web/treadmill/treadmill.py

```diff
@@ -0,0 +1,172 @@
+from bumble.device import Device, Advertisement, Connection, Peer
+from bumble.hci import Address, HCI_Reset_Command
+from bumble.core import AdvertisingData
+from bumble.gatt import (
+    GATT_HEART_RATE_SERVICE,
+    GATT_HEART_RATE_MEASUREMENT_CHARACTERISTIC,
+)
+from typing import Optional, Union
+
+from bumble.utils import AsyncRunner
+import asyncio
+import datetime
+
+
+# -----------------------------------------------------------------------------
+class Treadmill:
+    peer: Optional[Peer]
+
+    class ScanEntry:
+        def __init__(self, advertisement):
+            self.address = advertisement.address.to_string(False)
+            self.address_type = (
+                'Public',
+                'Random',
+                'Public Identity',
+                'Random Identity',
+            )[advertisement.address.address_type]
+            self.rssi = advertisement.rssi
+            self.data = advertisement.data.to_string('\n')
+
+    def __init__(self, hci_source, hci_sink):
+        super().__init__()
+        random_address = Address.generate_static_address()
+        self.device = Device.with_hci('Bumbleton', random_address, hci_source, hci_sink)
+        self.scan_entries = {}
+        self.listeners = {}
+        self.peer = None
+        self.device.on('advertisement', self.on_advertisement)
+        self.device.on('connection', self.on_connection)
+
+    async def start(self):
+        print('### Starting Scanner')
+        self.scan_entries = {}
+        self.emit_scanning_update()
+        await self.device.power_on()
+        await self.device.start_scanning()
+        print('### Scanner started')
+
+    async def stop(self):
+        # TODO: replace this once a proper reset is implemented in the lib.
+        await self.device.host.send_command(HCI_Reset_Command())
+        await self.device.power_off()
+        print('### Scanner stopped')
+
+    def emit_scanning_update(self):
+        if listener := self.listeners.get('scanning_updates'):
+            listener(list(self.scan_entries.values()))
+
+    def emit_hr_update(self, value: int, time: str):
+        if listener := self.listeners.get('hr_updates'):
+            listener({"value": value, "time": time})
+
+    def emit_connection_updates(self, connection: Connection):
+        if listener := self.listeners.get('connection_updates'):
+            listener(
+                {
+                    'handle': connection.handle,
+                    'role': connection.role_name,
+                    'self_address': str(connection.self_address),
+                    'peer_address': str(connection.peer_address),
+                    'is_encrypted': "Yes" if connection.is_encrypted else "No",
+                }
+            )
+
+    def emit_on_security_request(self):
+        if listener := self.listeners.get('on_security_request'):
+            listener()
+
+    def on(self, event_name, listener):
```

Comment:
You can just subclass `pyee.EventEmitter` and get all the listener management and event emission implemented already for you.